kernel/comps/softirq/src/lib.rs

```diff
@@ -119,25 +127,87 @@ static ENABLED_MASK: AtomicU8 = AtomicU8::new(0);
 
 cpu_local_cell! {
     static PENDING_MASK: u8 = 0;
+    static DISABLE_SOFTIRQ_COUNT: u8 = 0;
+}
+
+fn increase_disable_softirq_count() {
+    DISABLE_SOFTIRQ_COUNT.add_assign(1);
+}
+
+fn decrease_disable_softirq_count() {
+    DISABLE_SOFTIRQ_COUNT.sub_assign(1);
+}
+
+fn disable_softirq_count() -> u8 {
+    DISABLE_SOFTIRQ_COUNT.load()
+}
+
+#[clippy::has_significant_drop]
+#[must_use]
+struct DisableLocalSoftirqGuard(PhantomData<()>);
```

Comment:
I'm a bit confused as to what the benefit is of having both `DisableLocalSoftirqGuard` and `DisableBottomHalfGuard`.

Although I see that `DisableLocalSoftirqGuard` is an internal helper guard that never gets exposed, but its semantics are hard to make sense of without first disabling preemption.

I wander why not just call `{increase,decrease}_disable_softirq_count` directly in `DisableBottomHalfGuard`?